### ISSUE TITLE
Documentation clarification: action

### ISSUE THREAD

**Issue by chenl** [igordejanovic/parglare#66]
in `docs/actions/#built-in-actions` you write:

> `collect_optional` - Can be used for rules of the form 
> `Elements: Elements Element | Element | EMPTY;`
>Returns list;

Shouldn't that be `Elements: Elements Element | EMPTY;`?
This pattern is also in `docs/grammar_language#zero-or-more`

> `collect_sep_optional` - Can be used for rules of the form 
>  `Elements:  Elements separator Element | Element | EMPTY;`
> Returns list;

This will accept `<separator> <element>` which is probably not correct.
Note that this is already addressed in `docs/grammar_language/#zero-or-more_1`

**Comment by igordejanovic** [comment]
Yes, you are right. This actually should be corrected for the syntactic sugar transformation for `*` also, in both docs and implementation. Thanks for reporting.

**Comment by igordejanovic** [comment]
Nope, I was wrong. Syntactic sugar transformations seems correct. I think that `collect_optional` and `collect_sep_optional` should be removed as those are obviously wrong as you noticed and both should be implemented as choice of `collect` and `collect_sep` with `EMPTY` like it is done with [zero or more](http://www.igordejanovic.net/parglare/latest/grammar_language/#zero-or-more_1).

**Comment by chenl** [comment]
I don't sure if it needs documenting, but it wasn't obvious to me that

     Elements: Elements Element | EMPTY ;

and

    Elements: OneOrMoreElements | EMTPY ;
    OpneOrMoreElements: OneOrMoreElements Element ;

produce a different parse tree:

e.g.:

    [[[[[], 'a'], 'b'], 'c'], None]

vs

    [[['a', 'b'], 'c'], None]

**Comment by igordejanovic** [comment]
Your second example seems incomplete. How do you reduce first `OneOrMoreElements`?

**Comment by chenl** [comment]
Sorry fixed :point_up:

**Comment by igordejanovic** [comment]
It is pretty much similar to some classic text book examples.

First example:
```python
g = Grammar.from_string(r'''
Elements: Elements Element | EMPTY;
terminals
Element: /[abc]/;
''')

Parser(g).parse('a b c')  #  [[[[], 'a'], 'b'], 'c']

# or as as a tree
print(Parser(g, build_tree=True).parse('a b c').tree_str())
#  Elements[0->5]
#  Elements[0->3]
#    Elements[0->1]
#      Elements[0->0]    <- this is EMPTY reduction
#      Element[0->1, "a"]
#    Element[2->3, "b"]
#  Element[4->5, "c"]



```
Or if we draw that tree that would be something like:
![image](https://user-images.githubusercontent.com/1673425/47446306-428a0d80-d7bb-11e8-8dcd-0029e994e995.png)

Second example:
```python
g = Grammar.from_string(r'''
Elements: OneOrMoreE | EMPTY;
OneOrMoreE: OneOrMoreE Element | Element;
terminals
Element: /[abc]/;
''')

Parser(g).parse('a b c')  # [['a', 'b'], 'c']

# or as as a tree
print(Parser(g, build_tree=True).parse('a b c').tree_str())
# Elements[0->5]
#  OneOrMoreE[0->5]
#    OneOrMoreE[0->3]
#      OneOrMoreE[0->1]
#        Element[0->1, "a"]
#      Element[2->3, "b"]
#    Element[4->5, "c"]

```
Or if we draw that tree that would be something like:
![image](https://user-images.githubusercontent.com/1673425/47445636-b75c4800-d7b9-11e8-90bb-fd988afdf104.png)


I don't think that should need a better documentation as it behaves more or less
as described in the textbooks on the subject. Please tell me what confuses you here?

**Comment by igordejanovic** [comment]
Ah, ok I understand. I guess it is not documented but will need to check. The rule of the thumb is that each rule on the RHS will produce a node in the tree if the tree is built or an element in the list if nested lists are built (the default actions), or passed to action if custom action is registered.

**Comment by igordejanovic** [comment]
BTW, `Elems: Elems Elem | EMPTY ;`  is actually a `zero or more` as `EMPTY` is a valid reduction for `Elems`. The exact grammar pattern is usually related to the way you want reduction to be processed so there can be different ways to match zero or more but the order or reduction might differ. I don't think it is actually suggested that some form is preferred over the other. For example you can see that in the implementation of the `*` operator a [slightly different grammar pattern is used](http://www.igordejanovic.net/parglare/latest/grammar_language/#zero-or-more_1) to handle dealing with preferred shift strategy.

**Comment by igordejanovic** [comment]
Just a bit more clarification. Built-in actions in parglare are not different than any other user written action. You can see the implementation [here](https://github.com/igordejanovic/parglare/blob/master/parglare/actions.py), so in order to be used you must write your rule as required. This, of course, doesn't stops you from writing rule the way you like it but in that case you must supply your action. In any case, if you don't have any specific requirement on the way your reductions are performed on `zero/one - or -more` matches and you are writing the grammar from scratch then it is recommended to use [EBNF extensions](http://www.igordejanovic.net/parglare/latest/grammar_language/#syntactic-sugar-bnf-extensions) as they will make the grammar more concise and readable.
